Improvements

ArmorDarks opened the pull request:
Closes #3

A good portion of various improvements as discussed in #3.

Prettier and ESLint added as it is, without any fancy watchers. But this should be enough for this simple case.

Note, that `npm test` or `npm run lint` will fail right now, because I did not add ES2015 support right now into Gruntfile and test files. I wanted to know first are you ok with that. I didn't want to rewrite all that into older JS, which is significantly less readable in tests due to lack of arrow functions.

If you're ok with that, let me know, I will remain current codebase and add ES2015 support. Otherwise, I will rewrite tests...

As a further step, I'd also add ES2015 support into main task file and bump Node version alongside with it. Once again, it will significantly increase code readability.

Finally, I'd set Prettier to not include semicolons. They do not add anything to Grunt task but some good portion of the noise. Though, for some people, it's painful theme. Like tabs vs spaces, you know...

I also recommend adding Travis, since we have now `npm test` which he could use for every commit or PR. Quite handy staff.

ArmorDarks committed:
Add linting with ESLint and formatting with Prettier

ArmorDarks committed:
Enforce testing of the code before publishing

ArmorDarks committed:
Update license field to modern standard

ArmorDarks committed:
Add `.editorconfig`

ArmorDarks committed:
Format the code with Prettier

ArmorDarks committed:
Fix the typo

ArmorDarks committed:
Move requires outside of the plugin function for better performance

ArmorDarks committed:
Specify main entry point and files in `package.json`

ArmorDarks committed:
Add testing with Grunt

ArmorDarks committed:
Add fixtures for testing Grunt run

ArmorDarks committed:
Add Jest as a test runner

ArmorDarks committed:
Rely on Grunt `src` extensions types

ArmorDarks committed:
Do not limit extensions in Grunt testing tasks

ArmorDarks committed:
Add test for not an image case

ArmorDarks committed:
Reduce number of initial checks and exit earlier

ArmorDarks committed:
Do not check for path existence

Due to nature of the Grunt, received from the glob files paths always exist. Unless you're doing something very wrong...

ArmorDarks committed:
Do not check for `file.dest` existence second time, and that late

ArmorDarks committed:
Few minor code readability improvements

ArmorDarks committed:
Add message about number of processed files

ArmorDarks committed:
Specify version of `image-size`

ArmorDarks committed:
Do not check `objectConfig` property at the beginning since it doesn't help

ArmorDarks committed:
Add ability to process various data parts with `processName()` `processEntry()` and `processSizes()` options

ArmorDarks committed:
Allow to specify output file replacer and spacing

ArmorDarks committed:
Intended by default output file. Just for beauty sake

ArmorDarks committed:
Fix README

ArmorDarks committed:
Make `configObject` option `undefined` by default to reduce number of checks

ArmorDarks commented:
@saperio pinging just in case

saperio commented:
Hi! It take some time to review your PR. Thanks for your work!

ArmorDarks commented:
Sure, take your time :)

Though, now I start to think that I have to add PR with removed types check first. It is quite crucial for us because due to that check we can not work with SVG (they get filtered out). However, I didn't want to change code without having tests first anyway...

saperio commented on tasks/image_size.js:
```diff
@@ -7,53 +7,92 @@
+      });
+
+      if (!this.files.length) {
+        return grunt.log.error('No files or `options.configObject` specified.');
```
- as I understand, grunt.log.error returns Log object, so there is no sense to return it back
- 'No files specified.' is more correct messege here

ArmorDarks replied:
Returning log object is just a shortcut to exit, to avoid writing additional line.

In other words, to avoid writing this:

```js
if (!this.files.length) {
  grunt.log.error('No files or `options.configObject` specified.');
  return;
}
```

Technically it's doing the same thing. If it bothers you from the aesthetic point of view, I'll change it.

> 'No files specified.' is more correct messege here

Hm, it was present in the original script. Can't say why I didn't change it. Fixed.

saperio commented on tasks/image_size.js:
```diff
@@ -7,53 +7,92 @@
+      this.files.forEach(function(file) {
+        if (!file.src.length)
+          return grunt.log.error(
+            'No source files specified for ' + chalk.cyan(dest) + '.'
```
dest -> file.dest and it can be undefined (if configObject is specified)

ArmorDarks replied:
Right, I've completely missed it. Fixed

saperio replied:
But it still can be undefined, need additional check for that (and another message, maybe)

ArmorDarks replied:
Am I missing something? I'm checking here is it defined and the message has been updated accordingly: https://i.imgur.com/NGdmCAD.png

saperio replied:
If file.dest is undefined (and file.src array is empty), then message in https://github.com/ArmorDarks/grunt-image-size/blob/feature/prettier/tasks/image_size.js#L34 will be like this: 'No source files specified for undefined.'

ArmorDarks replied:
Right.

I've added handling for this case by simply checking `file.dest` _before_ `file.src`.

saperio commented on tasks/image_size.js:
```diff
@@ -7,53 +7,92 @@
+            'No source files specified for ' + chalk.cyan(dest) + '.'
+          );
+
+        if (!file.dest) return grunt.log.error('No dest file specified.');
```
- return Log object again
- if configObject is specified, there is no error here
- if both configObject and file.dest undefined - that is incorrect config (and you need to check it here and report with grunt.fail.warn)

ArmorDarks replied:
Yeap, missed that. Fixed in related to previous comment commit.

It will report this: https://i.imgur.com/A7ZsQK3.png

saperio commented on tasks/image_size.js:
```diff
@@ -7,53 +7,92 @@
+        file.src.forEach(function(src) {
+          if (!grunt.file.isFile(src)) return;
+
+          var size = sizer(src);
```
if src has non-supported format (or corrupted), sizer throws error, so it must be a try-catch block here

ArmorDarks replied:
Due to nature of Grunt, `dest` will be never incorrect or anything like that. But yeap, the file might end up being corrupt or of the unsupported extension.

Wrapping things into try-catch block sometimes makes code slower (since it prevents some important V8 optimization). But yeap, it seems to be the only solution here.

Generally, there are two ways:

1. Let it be like it is. The task will completely fail on fatal error. Well, not worst scenario.
2. Wrap it into try-catch.

   It makes sense only if we want to let the task run _till the end_, even if some of the files resulted in an error.

   Solution like this used, for instance, [here](https://github.com/ArmorDarks/grunt-gray-matter/blob/master/tasks/grayMatter.js)

   But I'm still questioning myself is it a reasonable thing to do, since it might be annoying in big projects. You have thousands of images, you run the task and only when it finished it says that few files resulted in an error. I wonder, wouldn't it be more reasonable to fail right away.

   And if we decide that failing right away is better — wrapping sizer into try-catch block doesn't add anything to it.

saperio replied:
> But I'm still questioning myself is it a reasonable thing to do, since it might be annoying in big projects
You right, but than we rely on to assumptions:
- grunt will catch all thrown errors
- image-size will throw always Error objects (and not just string, for example)
And I think, we can't. If in image-size code will be added this: ``throw 'some shit';``, grunt-image-size will be broken, because it end without error with this throw.

ArmorDarks replied:
Well, it isn't that simple with Grunt.

Just run few tests with errors, and you will see that during logging Grunt for some reason cuts the trace of the error: https://i.imgur.com/jVnyKQO.png

So, if the only purpose to provide a more meaningful error, by try-catching here we won't provide better context.

saperio replied:
No, the purpose  of try-catch here is to report at least some error. If you change ``throw new Error('test me');`` to ``throw 'test me';`` task will end up successfully.

saperio replied:
It strange, but if I put ``throw 'test';`` inside task, I've got this: https://i.imgur.com/y58mf3W.png

ArmorDarks replied:
Tested this and got the same result. More of it, if a naked throw is added, it magically makes pass `grunt image_size:noImage` without errors at all, which is definitely not right.

Seems like a bug or intentional strange design choice in Grunt. Normally scripts will fail on any throw. `Error` constructor just attaches to it trace object.

Well, that's not our issue, I'd say. I'd propose to consider that for future and maybe even open Grunt issue. But let's keep it out of the scope of this PR. If it will turn to be indeed an issue, we will add later try-catch as a workaround.

saperio replied:
Agree.

saperio commented on tasks/image_size.js:
```diff
@@ -7,53 +7,92 @@
+          grunt.config.set(options.configObject, sizes);
+        }
+
+        grunt.file.write(
```
you need to check, if file.dest is defined

ArmorDarks replied:
We're checking this earlier, to avoid running useless operations if task can't be accomplished anyway, here https://github.com/ArmorDarks/grunt-image-size/blob/7c445effedeeaa04ee98c97fd71acb5429825eeb/tasks/image_size.js#L36

saperio replied:
dest can be undefined and it is valid case if configObject defined (as in README: "outputs it to grunt Config Data and/or to dest file"), so we need to change condition in https://github.com/ArmorDarks/grunt-image-size/blob/7c445effedeeaa04ee98c97fd71acb5429825eeb/tasks/image_size.js#L36 and add it here

ArmorDarks replied:
Agree, fixed that.

saperio commented on tasks/image_size.js:
```diff
@@ -7,53 +7,92 @@
+          JSON.stringify(sizes, options.replacer, options.space)
+        );
+
+        grunt.log.writeln('File ' + chalk.cyan(file.dest) + ' created.');
```
I think grunt.verbose instead of grunt.log is more suitable here

ArmorDarks replied:
Logging in verbose won't help much, since in verbose mode `grunt.file.write` itself will log the same thing.

The purpose of that line is just to indicate in _normal mode_ that some dest files have been written. This subjective thing, but I always had feeling that it helps to see it even in normal mode, to ensure that everything works right.

saperio replied:
Ok

saperio commented on tasks/image_size.js:
```diff
@@ -7,53 +7,92 @@
+
+        grunt.file.write(
+          file.dest,
+          JSON.stringify(sizes, options.replacer, options.space)
```
What if processSizes returns string (or Buffer), with prepared output, so we do not need replacer and space options at all. And user not restricted with json format for output.

ArmorDarks replied:
Hm, good catch.

Right now task assumes that `processSizes` can return only serializable data, and nothing more.

We can make something like this:

```diff
      var options = this.options({
+       processSizes: function (sizes) {
+         return JSON.stringify(sizes, null, 2);
+       },
        replacer: null,
        space: 2
      });

...

-       if (typeof options.processSizes === 'function') {
          sizes = options.processSizes.call(file, sizes, file);
-       }

        if (shouldSetConfigObject) {
          grunt.config.set(options.configObject, sizes);
        }

        if (file.dest) {
          grunt.file.write(
            file.dest,
-           JSON.stringify(sizes, options.replacer, options.space)
+           sizes
          );

          grunt.log.writeln('File ' + chalk.cyan(file.dest) + ' created.');
        }

```

It means that `processSizes` now will have predefined value, and by default will stringify input into the file, but the user can override it.

Though, it is quite new for Grunt task configs, not a common thing. I have a feeling that it might make a configuration not that clear for newcomers.

saperio replied:
Agree, leave replacer and space options

ArmorDarks replied:
I generally like the idea that it might be possible to expand the output beyond serializable data. But yeap, let's leave it for the future when someones really would need it.

saperio commented:
And for test and gruntfile - keep them as is and add them to .eslintignore

ArmorDarks committed:
Fix `main` path

ArmorDarks commented on package.json:
```diff
@@ -15,18 +15,40 @@
-    {
-      "type": "MIT"
-    }
+  "main": "tasks/grayMatter.js",
```
That's regretful typo. Fixed.

ArmorDarks commented on package.json:
```diff
@@ -15,18 +15,40 @@
+  },
+  "scripts": {
+    "lint": "eslint \"./**/*.js\"",
+    "format": "eslint \"./**/*.js\" --fix",
```
I've mentioned it in PR description. I didn't do it because wasn't sure which path would you prefer.

My imho — it is way more practical to lint Gruntfile and tests too. So I'm adding node and es6 to ESlint env than.

ArmorDarks committed:
Specify Node and ES6 environment for ESLint

ArmorDarks committed:
Prettify scripts

ArmorDarks committed:
Do not declare options, which are `undefined` by default

ArmorDarks commented on tasks/image_size.js:
```diff
@@ -7,53 +7,92 @@
+        processName: undefined,
+        processEntry: undefined,
+        processSizes: undefined,
+        configObject: undefined,
```
Yeap, they aren't. I just was not sure do you prefer to keep them there or not.

ArmorDarks committed:
Fix error message

ArmorDarks commented:
Thanks for the review! Fixed some of the issues. I will return to other ones later.

saperio commented:
One more thing - configObject will be overwritten in multifile config format, but I fix this by myself.

ArmorDarks committed:
Fix `file.dest` typo

ArmorDarks committed:
Allow task to pass if no `file.dest` specified, but `options.configObject` declared

ArmorDarks committed:
Remove no longer required lib

ArmorDarks committed:
Add placeholder for `configObject` tests. They do not work right now. See `@todo` in test file

ArmorDarks commented:
Added more fixes and replied in few comments, which needs further discussion before I can finish them.

> One more thing - configObject will be overwritten in multifile config format, but I fix this by myself.

Yeap, that's something I've missed.

Though, the concept of `configObject` is generally has been unexpected to me. We never used anything like this, because Grunt loses context between runs, and usually writing to a file is the only way to keep context in Grunt. But I guess for your use case it makes sense...

Maybe the whole thing should be abstracted into the more high-level function. For instance, it can be something like:

```diff
      var options = this.options({
+       ouputFn: function (result, grunt) {
+         grunt.file.write(
+           file.dest,
+           JSON.stringify(sizes, options.replacer, options.space)
+         );
+       },
        replacer: null,
        space: 2
      });

...

-       if (shouldSetConfigObject) {
-         grunt.config.set(options.configObject, sizes);
-       }
-
-       if (file.dest) {
-         grunt.file.write(
-           file.dest,
-           JSON.stringify(sizes, options.replacer, options.space)
-         );
-
-         grunt.log.writeln('File ' + chalk.cyan(file.dest) + ' created.');
-       }
+ this.options.outputFn(sizes, grunt)
```

This will allow whenever needed to adjust `outputFn` to write into Grunt property or anything else.

But in general, it will make configuration less clear for newcomers.

saperio commented:
> Though, the concept of configObject is generally has been unexpected to me.

I use it to push data to next task in sequence. Back in days, it was static site generator on grunt, it read images and push data to dust, to generate html from template.

saperio commented:
This
```js
grunt.registerTask('generate_index', [
	'image_size',
	'process_image_data',
	'dusthtml:main'
]);
```

ArmorDarks commented:
> I use it to push data to next task in sequence. Back in days, it was static site generator on grunt, it read images and push data to dust, to generate html from template.

Yeap, I've got the idea how it is possible could be used. I just meant that it worked in very few scenarios, because Grunt loses context between runs, and it becomes a problem if you're using a watch mode.

Well, communication between tasks was always a problem in Grunt...

Anyway, as you've said, since it is a thing which you worked with, let's keep it beyond the scope of this PR. But I agree, it should be fixed.

Probably pulling `lodash/merge` would be the fastest way to do this... But be careful with mutations. Merging things between task runs sometimes can lead to very undesirable leaks during watch mode.

Btw, we will be using your library with our [static website generator](https://github.com/LotusTM/Kotsu) (coupled with the starter kit) too :) That's why I was so eager to update it...

I've replied to your comments above.

Besides them, the only thing that still demands discussion is Node version. I'd bump it to 4.0.0 and update code to use ES6. From my point of view, supporting Node 0.12 in 2018 is even a harmful thing to do :)

And along the way, I'd drop semicolons from JS. Less noise, more happiness.

saperio commented:
>  I'd bump it to 4.0.0 and update code to use ES6. From my point of view, supporting Node 0.12 in 2018 is even a harmful thing to do :)

Yeap.

> And along the way, I'd drop semicolons from JS. Less noise, more happiness.

Code is naked without semicolons, I can't accept that :)

ArmorDarks committed:
Check is `file.dest` or `options.configObject` specified before checking `file.src` to avoid error message with undefined

ArmorDarks committed:
Mention for which property can't be found source files, if there any specified

ArmorDarks committed:
Handle cases, when `file.src` might end up completely undefined

ArmorDarks committed:
Unify brackets after single-line `if`

ArmorDarks committed:
Bump min required Node version to 6.0.0

ArmorDarks committed:
Use ES6 in task file

ArmorDarks commented:
Once again replied to comments and fixed few things.

Also, bumped Node version up to 6.0.0 (I wrongly stated 4.0.0. It would be better to require at least 6.0.0 which is more secure).

And added ES6 to task file, as we agreed.

> Code is naked without semicolons, I can't accept that :)

Kinda spaces vs tabs topic. From my point of view, it is not a matter of how it looks, but a matter of pragmatism. Lack of semicolons just makes writing code faster, reduces stupid errors with forgotten semicolons and in general makes less unnecessary noise.

But it is your library, so yours are rules. I'm more concerned about performance :)

ArmorDarks committed:
Unify README code snippets indentation

ArmorDarks committed:
Update Development and testing section

ArmorDarks commented:
Nice, thanks!

You also might want to add Travis now.

Here is example file: https://github.com/ArmorDarks/grunt-gray-matter/blob/master/.travis.yml

Travis-side configuration is trivial and takes few clicks. Or you might want to use CircleCI, which is much faster and free too.
